dbms/src/Storages/Page/PageStorage.h

```diff
@@ -317,77 +336,540 @@ class PageStorage : private boost::noncopyable
 private:
+    PageStorage::SnapshotPtr toConcreteV3Snapshot() const
+    {
+        return snap ? toConcreteMixedSnapshot(snap)->getV3Snasphot() : snap;
+    }
+
+    PageStorage::SnapshotPtr toConcreteV2Snapshot() const
+    {
+        return snap ? toConcreteMixedSnapshot(snap)->getV2Snasphot() : snap;
+    }
+
+private:
+    const PageStorageRunMode run_mode;
     NamespaceId ns_id;
-    PageStoragePtr storage;
+    PageStoragePtr storage_v2;
+    PageStoragePtr storage_v3;
     PageStorage::SnapshotPtr snap;
     ReadLimiterPtr read_limiter;
 };
+using PageReaderPtr = std::shared_ptr<PageReader>;
+
+class PageWriter : private boost::noncopyable
+{
+public:
+    PageWriter(PageStorageRunMode run_mode_, PageStoragePtr storage_v2_, PageStoragePtr storage_v3_)
+        : run_mode(run_mode_)
+        , storage_v2(storage_v2_)
+        , storage_v3(storage_v3_)
+    {
+    }
+
+    void write(WriteBatch && write_batch, WriteLimiterPtr write_limiter) const
+    {
+        switch (run_mode)
+        {
+        case PageStorageRunMode::ONLY_V2:
+        {
+            storage_v2->write(std::move(write_batch), write_limiter);
+            break;
+        }
+        case PageStorageRunMode::ONLY_V3:
+        {
+            storage_v3->write(std::move(write_batch), write_limiter);
+            break;
+        }
+        case PageStorageRunMode::MIX_MODE:
+        {
+            const auto & ns_id = write_batch.getNamespaceId();
+            WriteBatch wb_for_v2{ns_id};
+            WriteBatch wb_for_put_v3{ns_id};
+
+            for (const auto & write : write_batch.getWrites())
+            {
+                switch (write.type)
+                {
+                // PUT/PUT_EXTERNAL only for V3
+                case WriteBatch::WriteType::PUT:
+                case WriteBatch::WriteType::PUT_EXTERNAL:
+                {
+                    break;
+                }
+                // Both need del in v2 and v3
+                case WriteBatch::WriteType::DEL:
+                {
+                    wb_for_v2.copyWrite(write);
+                    break;
+                }
+                case WriteBatch::WriteType::REF:
+                {
+                    PageId resolved_page_id = storage_v3->getNormalPageId(ns_id,
+                                                                          write.ori_page_id,
+                                                                          /*snapshot*/ nullptr,
+                                                                          false);
+                    // if normal id is not ok, read from v2 and create a new put + ref
+                    if (resolved_page_id == INVALID_PAGE_ID)
+                    {
+                        const auto & entry_for_put = storage_v2->getEntry(ns_id, write.ori_page_id, /*snapshot*/ {});
+                        if (entry_for_put.file_id != 0)
+                        {
+                            auto page_for_put = storage_v2->read(ns_id, write.ori_page_id);
```

Comment:
The `page_for_put` is just a local var here, when we call `storage_v3->write(std::move(wb_for_put_v3), write_limiter)` can we still access to the data?

Reply:
build with asan
```
[ RUN      ] PageStorageMixedTest.Ref
...
=================================================================
==202679==ERROR: AddressSanitizer: heap-use-after-free on address 0x61900025a380 at pc 0x00000361f0b7 bp 0x7ffd08ff0a70 sp 0x7ffd08ff0238
READ of size 1024 at 0x61900025a380 thread T0
    #0 0x361f0b6 in __asan_memcpy /root/llvm-project/compiler-rt/lib/asan/asan_interceptors_memintrinsics.cpp:22:3
    #1 0x37a7a65 in DB::ReadBuffer::read(char*, unsigned long) /DATA/disk1/jaysonhuang/tics/dbms/src/IO/ReadBuffer.h:149:13
    #2 0x37a7a65 in DB::ReadBuffer::readStrict(char*, unsigned long) /DATA/disk1/jaysonhuang/tics/dbms/src/IO/ReadBuffer.h:160:18
    #3 0x19e1d8f9 in DB::PS::V3::BlobStore::write(DB::WriteBatch&, std::__1::shared_ptr<DB::WriteLimiter> const&) /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/V3/BlobStore.cpp:178:32
    #4 0x19e836d0 in DB::PS::V3::PageStorageImpl::writeImpl(DB::WriteBatch&&, std::__1::shared_ptr<DB::WriteLimiter> const&) /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/V3/PageStorageImpl.cpp:105:28
    #5 0x6d918d4 in DB::PageStorage::write(DB::WriteBatch&&, std::__1::shared_ptr<DB::WriteLimiter> const&) /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/PageStorage.h:248:9
    #6 0x6d918d4 in DB::PageWriter::write(DB::WriteBatch&&, std::__1::shared_ptr<DB::WriteLimiter>) const /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/PageStorage.h:778:29
    #7 0x7548e2a in DB::PS::V3::tests::PageStorageMixedTest_Ref_Test::TestBody() /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/V3/tests/gtest_page_storage_mix_mode.cpp:449:26
    #8 0x13e0e2f0 in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2401:10
    #9 0x13e0e2f0 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2437:14
    #10 0x13dbe4d8 in testing::Test::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2473:5
    #11 0x13dc15e7 in testing::TestInfo::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2655:11
    #12 0x13dc2b4f in testing::TestCase::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2773:28
    #13 0x13dddd48 in testing::internal::UnitTestImpl::RunAllTests() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:4673:43
    #14 0x13e10320 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2401:10
    #15 0x13e10320 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2437:14
    #16 0x13ddc96a in testing::UnitTest::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:4281:10
    #17 0x795863a in RUN_ALL_TESTS() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/include/gtest/gtest.h:2237:46
    #18 0x795863a in main /DATA/disk1/jaysonhuang/tics/dbms/src/TestUtils/gtests_dbms_main.cpp:36:16
    #19 0x7f92ba9a8554 in __libc_start_main (/lib64/libc.so.6+0x22554)
    #20 0x35a3768 in _start (/DATA/disk1/jaysonhuang/tics/cmake-build-asan/dbms/gtests_dbms+0x35a3768)

0x61900025a380 is located 0 bytes inside of 1024-byte region [0x61900025a380,0x61900025a780)
freed by thread T0 here:
    #0 0x361fa92 in free /root/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:111:3
    #1 0x1bc1826b in Allocator<false>::free(void*, unsigned long) /DATA/disk1/jaysonhuang/tics/dbms/src/Common/Allocator.cpp:127:9
    #2 0x72d0e21 in std::__1::__function::__value_func<void (char*)>::operator()(char*&&) const /opt/tiflash-env/sysroot/bin/../include/c++/v1/__functional/function.h:498:16
    #3 0x72d0e21 in std::__1::function<void (char*)>::operator()(char*) const /opt/tiflash-env/sysroot/bin/../include/c++/v1/__functional/function.h:1175:12
    #4 0x72d0e21 in std::__1::__shared_ptr_pointer<char*, std::__1::function<void (char*)>, std::__1::allocator<char> >::__on_zero_shared() /opt/tiflash-env/sysroot/bin/../include/c++/v1/__memory/shared_ptr.h:267:5
    #5 0x6d9168a in std::__1::__shared_count::__release_shared() /opt/tiflash-env/sysroot/bin/../include/c++/v1/__memory/shared_ptr.h:177:9
    #6 0x6d9168a in std::__1::__shared_weak_count::__release_shared() /opt/tiflash-env/sysroot/bin/../include/c++/v1/__memory/shared_ptr.h:219:27
    #7 0x6d9168a in std::__1::shared_ptr<char>::~shared_ptr() /opt/tiflash-env/sysroot/bin/../include/c++/v1/__memory/shared_ptr.h:959:19
    #8 0x6d9168a in DB::Page::~Page() /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/Page.h:35:8
    #9 0x6d9168a in DB::PageWriter::write(DB::WriteBatch&&, std::__1::shared_ptr<DB::WriteLimiter>) const /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/PageStorage.h:754:25
    #10 0x7548e2a in DB::PS::V3::tests::PageStorageMixedTest_Ref_Test::TestBody() /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/V3/tests/gtest_page_storage_mix_mode.cpp:449:26
    #11 0x13e0e2f0 in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2401:10
    #12 0x13e0e2f0 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2437:14
    #13 0x13dbe4d8 in testing::Test::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2473:5
    #14 0x13dc15e7 in testing::TestInfo::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2655:11
    #15 0x13dc2b4f in testing::TestCase::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2773:28
    #16 0x13dddd48 in testing::internal::UnitTestImpl::RunAllTests() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:4673:43
    #17 0x13e10320 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2401:10
    #18 0x13e10320 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2437:14
    #19 0x13ddc96a in testing::UnitTest::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:4281:10
    #20 0x795863a in RUN_ALL_TESTS() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/include/gtest/gtest.h:2237:46
    #21 0x795863a in main /DATA/disk1/jaysonhuang/tics/dbms/src/TestUtils/gtests_dbms_main.cpp:36:16
    #22 0x7f92ba9a8554 in __libc_start_main (/lib64/libc.so.6+0x22554)

previously allocated by thread T0 here:
    #0 0x361fcfd in malloc /root/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:129:3
    #1 0x1bc17384 in Allocator<false>::alloc(unsigned long, unsigned long) /DATA/disk1/jaysonhuang/tics/dbms/src/Common/Allocator.cpp:94:23
    #2 0x19d47885 in DB::PS::V2::PageFile::Reader::read(std::__1::vector<std::__1::pair<unsigned long, DB::PageEntry>, std::__1::allocator<std::__1::pair<unsigned long, DB::PageEntry> > >&, std::__1::shared_ptr<DB::ReadLimiter> const&, bool) /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/V2/PageFile.cpp:892:43
    #3 0x19d840d1 in DB::PS::V2::PageStorage::readImpl(unsigned long, unsigned long, std::__1::shared_ptr<DB::ReadLimiter> const&, std::__1::shared_ptr<DB::PageStorageSnapshot>, bool) /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/V2/PageStorage.cpp:645:25
    #4 0x6d90196 in DB::PageStorage::read(unsigned long, unsigned long, std::__1::shared_ptr<DB::ReadLimiter> const&, std::__1::shared_ptr<DB::PageStorageSnapshot>, bool) /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/PageStorage.h:260:16
    #5 0x6d90196 in DB::PageWriter::write(DB::WriteBatch&&, std::__1::shared_ptr<DB::WriteLimiter>) const /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/PageStorage.h:732:61
    #6 0x7548e2a in DB::PS::V3::tests::PageStorageMixedTest_Ref_Test::TestBody() /DATA/disk1/jaysonhuang/tics/dbms/src/Storages/Page/V3/tests/gtest_page_storage_mix_mode.cpp:449:26
    #7 0x13e0e2f0 in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2401:10
    #8 0x13e0e2f0 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2437:14
    #9 0x13dbe4d8 in testing::Test::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2473:5
    #10 0x13dc15e7 in testing::TestInfo::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2655:11
    #11 0x13dc2b4f in testing::TestCase::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2773:28
    #12 0x13dddd48 in testing::internal::UnitTestImpl::RunAllTests() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:4673:43
    #13 0x13e10320 in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2401:10
    #14 0x13e10320 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:2437:14
    #15 0x13ddc96a in testing::UnitTest::Run() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/src/gtest.cc:4281:10
    #16 0x795863a in RUN_ALL_TESTS() /DATA/disk1/jaysonhuang/tics/contrib/googletest/googletest/include/gtest/gtest.h:2237:46
    #17 0x795863a in main /DATA/disk1/jaysonhuang/tics/dbms/src/TestUtils/gtests_dbms_main.cpp:36:16
    #18 0x7f92ba9a8554 in __libc_start_main (/lib64/libc.so.6+0x22554)

SUMMARY: AddressSanitizer: heap-use-after-free /root/llvm-project/compiler-rt/lib/asan/asan_interceptors_memintrinsics.cpp:22:3 in __asan_memcpy
Shadow bytes around the buggy address:
  0x0c3280043420: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c3280043430: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c3280043440: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c3280043450: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fa fa
  0x0c3280043460: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c3280043470:[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3280043480: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c3280043490: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c32800434a0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c32800434b0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c32800434c0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==202679==ABORTING
```